### ISSUE TITLE
vmagent still scrapes target after pod was deleted

### ISSUE THREAD

**Issue by okzheng** [VictoriaMetrics/VictoriaMetrics#3249]
We have same targets scraped by Prometheus and vmagent. Targets have been configured in kubernetes_sd_configs. Sometimes when a pod( a target) was deleted from kubernetes cluster, the vmagent still scrapes from it. The scrape does not stop until a few scrape intervals have elapsed.
The configuration
```
vmagent version: 1.79.2
scrape interval: 30s

- job_name: serviceScrape/prod/api-gateway/0
  metrics_path: /metrics
  relabel_configs:
  - source_labels: [__meta_kubernetes_service_label_app]
    regex: api-gateway
    action: keep
  - source_labels: [__meta_kubernetes_endpoint_port_name]
    regex: http-metrics
    action: keep
  - source_labels: [__meta_kubernetes_endpoint_address_target_kind, __meta_kubernetes_endpoint_address_target_name]
    separator: ;
    target_label: node
    regex: Node;(.*)
    replacement: ${1}
  - source_labels: [__meta_kubernetes_endpoint_address_target_kind, __meta_kubernetes_endpoint_address_target_name]
    separator: ;
    target_label: pod
    regex: Pod;(.*)
    replacement: ${1}
  - source_labels: [__meta_kubernetes_pod_name]
    target_label: pod
  - source_labels: [__meta_kubernetes_pod_container_name]
    target_label: container
  - source_labels: [__meta_kubernetes_namespace]
    target_label: namespace
  - source_labels: [__meta_kubernetes_service_name]
    target_label: service
  - source_labels: [__meta_kubernetes_pod_label_app]
    target_label: app
    regex: (.+)
    replacement: ${1}
  - source_labels: [__meta_kubernetes_pod_label_team]
    target_label: team
    regex: (.+)
    replacement: ${1}
  - source_labels: [__meta_kubernetes_pod_label_owner]
    target_label: owner
    regex: (.+)
    replacement: ${1}
  - source_labels: [__meta_kubernetes_service_name]
    target_label: job
    replacement: ${1}
  - target_label: endpoint
    replacement: http-metrics
  sample_limit: 10000
  kubernetes_sd_configs:
  - role: endpoints
    kubeconfig_file: ""
    namespaces:
      own_namespace: false
      names:
      - prod
```

a real example:
```
2022-10-17T06:44:06.775237Z replicaset-controller starts to delete the pod api-gateway-6cdfc8d675-m4qtt

2022-10-17T06:44:34.288686Z kubelet deleted the pod api-gateway-6cdfc8d675-m4qtt

2022-10-17T06:44:36.720Z vmagent logging a scrape error. cannot scrape "http://172.16.162.46:9090/metrics" ......: error
when scraping "http://172.16.162.46:9090/metrics": dial tcp4 172.16.162.46:9090: i/o timeout; try -enableTCP6 command-line flag if you scrape ipv6 addresses

2022-10-17T06:45:04.749Z same srape error log

2022-10-17T06:45:32.001Z srape error log

2022-10-17T06:46:02.017Z srape error log

2022-10-17T06:46:09.173Z        info    VictoriaMetrics/lib/promscrape/discovery/kubernetes/api_watcher.go:719  kubernetes sd update object prod/api-gateway-6cdfc8d675-m4qtt, uw https://172.17.240.1:443/api/v1/namespaces/prod/pods

2022-10-17T06:46:32.033Z srape error log

2022-10-17T06:47:01.986Z srape error log

2022-10-17T06:47:27.992Z        info    VictoriaMetrics/lib/promscrape/discovery/kubernetes/api_watcher.go:719  kubernetes sd update object prod/api-gateway-6cdfc8d675-m4qtt, uw https://172.17.240.1:443/api/v1/namespaces/prod/pods

2022-10-17T06:47:29.438Z        info    VictoriaMetrics/lib/promscrape/discovery/kubernetes/api_watcher.go:719  kubernetes sd update object prod/api-gateway-6cdfc8d675-m4qtt, uw https://172.17.240.1:443/api/v1/namespaces/prod/pods

2022-10-17T06:47:32.002Z srape error log

2022-10-17T06:47:40.224Z        info    VictoriaMetrics/lib/promscrape/scraper.go:371   close deleted scraper, ......

2022-10-17T06:47:41.493Z        info    VictoriaMetrics/lib/promscrape/discovery/kubernetes/api_watcher.go:719  kubernetes sd update object prod/api-gateway-6cdfc8d675-m4qtt, uw https://172.17.240.1:443/api/v1/namespaces/prod/pods

2022-10-17T06:47:41.493Z        info    VictoriaMetrics/lib/promscrape/discovery/kubernetes/api_watcher.go:729  kubernetes sd deleted object prod/api-gateway-6cdfc8d675-m4qtt, uw https://172.17.240.1:443/api/v1/namespaces/prod/pods
```
I just add simple logs in a few functions for debugging.
```
// lib/promscrape/discovery/kubernetes/api_watcher.go
switch we.Type {
case "ADDED", "MODIFIED":
	o, err := uw.parseObject(we.Object)
	if err != nil {
		return fmt.Errorf("cannot parse %s object: %w", we.Type, err)
	}
	key := o.key()
	uw.gw.mu.Lock()
	uw.updateObjectLocked(key, o)
	logger.Infof("kubernetes sd update object %s, uw %s", key, uw.apiURL) //
	uw.gw.mu.Unlock()
case "DELETED":
	o, err := uw.parseObject(we.Object)
	if err != nil {
		return fmt.Errorf("cannot parse %s object: %w", we.Type, err)
	}
	key := o.key()
	uw.gw.mu.Lock()
	uw.removeObjectLocked(key)
	logger.Infof("kubernetes sd deleted object %s, uw %s", key, uw.apiURL) //
	uw.gw.mu.Unlock()
...}
```

```
// lib/promscrape/scraper.go
for key, sc := range sg.m {
if _, ok := swsMap[key]; !ok {
	close(sc.stopCh) 
	logger.Infof("close deleted scraper, key: %s", key) //
	stoppedChs = append(stoppedChs, sc.stoppedCh)
	delete(sg.m, key)
	deletionsCount++
}
}
```

In contrast, Prometheus stoped scraping in a short time, when the pod was deleted. This should be expected.
I can't find the reason exactly and a resolution. 
In addition, I wonder why kubernetes sd uses raw http instead use client-go. 
Thanks.

**Comment by denisgolius** [comment]
Hi @okzheng !
You see this behavior because `-promscrape.kubernetesSDCheckInterval` is set to 30 sec by default, see https://docs.victoriametrics.com/#list-of-command-line-flags
```
-promscrape.kubernetesSDCheckInterval duration
     Interval for checking for changes in Kubernetes API server. This works only if kubernetes_sd_configs is configured in '-promscrape.config' file. See https://docs.victoriametrics.com/sd_configs.html#kubernetes_sd_configs for details (default 30s)
```
Feel free to set the value up to your needs.

**Comment by okzheng** [comment]
Sorry, the flag did not work.  I have captured and analysed network packages which was from k8s apiserver to vmagent. I found that after several minutes pod was deleted, apiserver send event MODIFIED and DELETED about the pod to vmagent. Sometimes more than ten minutes. Then the pod was removed from scrape works. And connections established between vmagent and apiserver is more than 40, but between prometheus with same scrape configuration and apiserver is only 1.
According to source code of vmagent, vmagent would establish 3 connections to apiserver for watching pod, service, endpoints in each namespace. When I configured to scrape targets from a single namespace, it was indeed 3 connections. Then the problem above is disappeared.
So I'm not sure if too many connections may cause the problems. Or others?

**Comment by valyala** [comment]
@okzheng , could you verify whether `vmagent` has enough free CPU and network bandwidth for the workload it processes? These params can be monitored via the official Grafana dashboard for vmagent according to [these docs](https://docs.victoriametrics.com/vmagent.html#monitoring). If it has no enough CPU and/or network bandwidth resources, then the discovery of new/updated objects from Kubernetes apiserver may slow down significantly. Try also upgrading [vmagent](https://docs.victoriametrics.com/vmagent.html) to the latest release - [v1.83.0 at the moment](https://github.com/VictoriaMetrics/VictoriaMetrics/releases) - it contains some optimizations, which may reduce its CPU usage a bit.

> connections established between vmagent and apiserver is more than 40, but between prometheus with same scrape configuration and apiserver is only 1.
According to source code of vmagent, vmagent would establish 3 connections to apiserver for watching pod, service, endpoints in each namespace. When I configured to scrape targets from a single namespace, it was indeed 3 connections. Then the problem above is disappeared.

Thanks for the pointer - we'll try reducing the number of established connections from vmagent to Kubernetes apiserver when many scrape configs contain different filters / namespaces.

**Comment by TomGudman** [comment]
Hi,

We observe a similar issue. CPU load is null. CPU has no limits in place.
vmagent (we run latest) keep scraping and eventually it will recover but it takes a while (> 15min).
We can trigger the vmagent alert "fails to scrape targets for last 15min" by restarting lots of pods, making this default monitoring pointless and very noisy on containers (re)deploys. This alert has value to detect real issue IF vmagent could detect that the target is deleted and refresh faster.

@okzheng  do you still experience this issue or did you find a workaround?

**Comment by okzheng** [comment]
I have not observed high CPU load yet. But the issue always exists with current version v1.84.0. @TomGudman

**Comment by maxtacu** [comment]
+1 looking for solution. We use a lot of spot instances in our k8s cluster and have constantly triggered alerts "fails to scrape targets for last 15m" for pods that are not existing anymore. Looking forward to the solution.

**Comment by Noamshmueli** [comment]
+1

**Comment by zekker6** [comment]
This issue was fixed in [v1.97.0](https://docs.victoriametrics.com/changelog/#v1970):
> BUGFIX: [vmagent](https://docs.victoriametrics.com/vmagent/): do not discover scrape targets for already terminated pods and containers in [kubernetes_sd_configs](https://docs.victoriametrics.com/sd_configs/#kubernetes_sd_configs). Such pods and containers cannot be scraped and cannot resurrect, so there is no sense in generating scrape targets for them.

Closing this as completed. Feel free to reopen if the issue still occurs on versions newer than v1.97.0.
